### ISSUE TITLE
Add hierarchical sparse indexer for bounded encoder-memory retrieval

### ISSUE THREAD

**Issue by iamdarshg** [iamdarshg/KestrelVL-7B#6]
Parent epic: #2
Depends on: #4, #5
Related long-context runtime: #1

## Goal

Implement a two-stage hierarchical sparse indexer so decoder layers can retrieve useful encoder-memory locations without repeatedly searching the entire prompt context.

## Target retrieval flow

```text
encoder memory (potentially very long)
        |
        v
coarse index / candidate generator
        |
    bounded candidate pool
   target ~2K entries initially
        |
        v
fine index / scorer
        |
   Top-K ~256–512
        |
        v
CSA2 Full/Reindex/Reuse consumers
```

All sizes must remain configurable and evidence-driven.

## Requirements

- A Full-mode layer may generate a broad candidate pool.
- Reindex layers search only the candidate pool, not the full context.
- Reuse layers consume prior Top-K indices without re-searching.
- Support both detail-memory and semantic-memory sources from #4.
- Deterministic Top-K and tie-breaking.
- BF16 training projections.
- INT8 stored index-state deployment mode plus BF16 quality mode, consistent with #1.
- Chunked candidate scoring; never materialize a dense full-context score tensor when avoidable.
- Explicit telemetry for:
  - candidate-pool size;
  - Top-K;
  - index bytes/token;
  - retrieval latency;
  - recall against dense/reference retrieval;
  - fraction of decoder time spent indexing.

## Training targets

Support supervision/distillation from:

- teacher attention relevance;
- teacher-selected prompt positions;
- downstream logit/hidden-state reconstruction;
- synthetic retrieval tasks with known supporting spans.

## Tests

- exact agreement with a reference implementation on tiny inputs;
- deterministic candidate generation and Top-K;
- candidate-bound enforcement in Reindex mode;
- no full-context scan in Reuse mode;
- long-context memory scaling sweep;
- far-context retrieval test;
- state serialization/resume.

## Acceptance criteria

- [ ] Later indexing cost is bounded by candidate-pool size rather than raw context length.
- [ ] Top-K retrieval quality is measured against a dense/reference scorer.
- [ ] INT8 and BF16 index-state modes both work.
- [ ] No dense `T x T` allocation is introduced.
- [ ] Index state integrates with #1 bounded-cache/state interfaces.
- [ ] CSA2 Reindex/Reuse layers can consume the indexer without hidden full-context work.

Reference: https://huggingface.co/deepseek-ai/DeepSeek-V4.1-Flash/blob/main/README.md

**Comment by iamdarshg** [comment]
## Batch 2 CPU evidence — core criteria PASS (tiny doubles)

**Commit:** \15ef3ae\ (\src/model/sparse_indexer.py\). **Tests:** \	ests/test_sparse_indexer.py\ — 15/15 pass on CPU (~4.5s). Full suite 174 green.

**Proof:**
- Bounded cost: coarse scores S in chunks, fine scores C ≤ pool only; \positions_coarse_scored\/\positions_fine_scored\ telemetry; no [Q,S] tensor outside labeled \dense_reference_topk\ (tested via accounting).
- Quality vs dense: \
ecall_vs_dense = 1.0\ on tiny inputs (auto-computed ≤8192 positions); \
ecall_at_k\ correct (1.0 identical / 0.0 disjoint / partial; tested). \synthetic_copy_task\ + \span_recall\ achieve 1.0 on known spans (tested).
- INT8 + BF16: both modes run; measured \index_bytes_per_token\ int8 (12B w/ scale+zp) < bf16 (16B) at coarse_dim 8 (tested).
- Determinism: stable double-argsort Top-K, no randomness/dropout anywhere (tested twice-identical).
- Sources: semantic + detail both work and are recorded in telemetry (tested). Pure content dot-product × learned temperature, no positional bias (documented for far-context).
- Resume: snapshot/metadata/strict-load round-trip; structural mismatch raises; \chunk_size\ deliberately exempt so resume works across chunkings (tested).

**Remaining:** long-context scaling sweep + far-context retrieval on real memory — needs GPU run (L4 spot approval pending).  spent.
